internal/proxy/cache.go

```diff
@@ -0,0 +1,203 @@
+		defaultOn:    cc.DefaultEnabled,
+		requestLocks: locksutil.CreateLocks(),
+	}
+}
+
+// constructs the CacheKey for this request and token and returns the SHA256
+// hash
+func makeRequestHash(logger *log.Logger, r *http.Request, token string) (string, error) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		if r.Body != nil {
+			if err := r.Body.Close(); err != nil {
+				logger.Printf("error closing request body: %s", err)
+			}
+		}
+		return "", fmt.Errorf("failed to read request body: %w", err)
+	}
+	if r.Body != nil {
+		if err := r.Body.Close(); err != nil {
+			logger.Printf("error closing request body: %s", err)
+		}
+	}
+	r.Body = ioutil.NopCloser(bytes.NewReader(reqBody))
+	cacheKey := &CacheKey{
+		Token:       token,
+		Request:     r,
+		RequestBody: reqBody,
+	}
+
+	cacheKeyHash, err := computeRequestID(cacheKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to compute request hash")
+	}
+	return cacheKeyHash, nil
+}
+
+// computeRequestID results in a value that uniquely identifies a request
+// received by the proxy. It does so by SHA256 hashing the serialized request
+// object containing the request path, query parameters and body parameters.
+func computeRequestID(key *CacheKey) (string, error) {
+	var b bytes.Buffer
+
+	if key == nil || key.Request == nil {
+		return "", fmt.Errorf("cache key is nil")
+	}
+
+	cloned := key.Request.Clone(context.Background())
+	cloned.Header.Del(api.HeaderIndex)
+	cloned.Header.Del(VaultForwardHeaderName)
+	cloned.Header.Del(VaultInconsistentHeaderName)
+	cloned.Header.Del(VaultCacheControlHeaderName)
+	// Serialize the request
+	if err := cloned.Write(&b); err != nil {
+		return "", fmt.Errorf("failed to serialize request: %v", err)
+	}
+
+	// Reset the request body after it has been closed by Write
+	key.Request.Body = ioutil.NopCloser(bytes.NewReader(key.RequestBody))
+
+	// Append key.Token into the byte slice. Just in case the token was only
+	// passed directly in CacheKey.Token, and not in a header.
+	if _, err := b.Write([]byte(key.Token)); err != nil {
+		return "", fmt.Errorf("failed to write token to hash input: %w", err)
+	}
+
+	return hex.EncodeToString(cryptoutil.Blake2b256Hash(b.String())), nil
+}
+
+func (c *Cache) Set(keyStr string, data *CacheData) {
+	c.data.Set(keyStr, data, gocache.DefaultExpiration)
+}
+
+func (c *Cache) Get(keyStr string) (data *CacheData, err error) {
+	dataRaw, found := c.data.Get(keyStr)
+	if found && dataRaw != nil {
+		var ok bool
+		data, ok = dataRaw.(*CacheData)
+		if !ok {
+			return nil, fmt.Errorf("failed to convert cache item to CacheData for key %v", keyStr)
+		}
+	}
+
+	return data, nil
+}
+
+func (c *Cache) Remove(keyStr string) {
+	c.data.Delete(keyStr)
+}
+
+func setupCache(cacheConfig config.CacheConfig) *Cache {
+	if cacheConfig.TTL <= 0 {
+		return nil
+	}
+	return NewCache(cacheConfig)
+}
+
+func parseCacheOptions(cacheControlHeader string) *CacheOptions {
+	values := strings.Split(cacheControlHeader, ",")
```

Comment:
If we support passing in more that one caching option (with the most restrictive one winning), then we should also parse for the case we're `X-Vault-Cache-Control` is passed more than once since that's [equivalent](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2) to passing a comma-separated list

Reply:
Yep, that's a good point, added the extra parsing in 675aa462eade8a80299d379d8b89a097d19de39f.